In [91]:
import os
import sys
import pandas as pd
from sklearn.linear_model import LogisticRegression
import pickle
sys.path.append(os.path.abspath('../Scripts'))
import warnings
warnings.filterwarnings('ignore')

In [92]:
import eda as ed

In [93]:
import Model as m

In [94]:
df = 'features_binned.csv'
features = ed.load_data(df)

2024-10-12 02:45:15,561 - INFO - Attempting to load data from features_binned.csv
2024-10-12 02:45:15,561 - INFO - Attempting to load data from features_binned.csv
2024-10-12 02:45:15,593 - INFO - Successfully loaded data from features_binned.csv
2024-10-12 02:45:15,593 - INFO - Successfully loaded data from features_binned.csv


In [95]:
features

CustomerId  Transaction_Frequency  Avg_Transaction_Amount  \
0        CustomerId_1               0.000000                0.045974   
1       CustomerId_10               0.000000                0.045974   
2     CustomerId_1001               0.000978                0.047525   
3     CustomerId_1002               0.002445                0.047124   
4     CustomerId_1003               0.001222                0.047451   
...               ...                    ...                     ...   
3737   CustomerId_992               0.001222                0.047451   
3738   CustomerId_993               0.000978                0.047525   
3739   CustomerId_994               0.024450                0.047678   
3740   CustomerId_996               0.003912                0.047988   
3741   CustomerId_998               0.005134                0.047585   

      Total_Transaction_Volume  Transaction_Recency  \
0                     0.556885             0.923077   
1                     0.556885             0.923077   
2                     0.557044             0.978022   
3                     0.556961             0.285714   
4                     0.557044             0.131868   
...                        ...                  ...   
3737                  0.557044             0.054945   
3738                  0.557044             0.285714   
3739                  0.559826             0.010989   
3740                  0.557676             0.747253   
3741                  0.557469             0.000000   

      Transaction_Amount_Variability  Avg_Transaction_Hour  \
0                           0.005076              0.695652   
1                           0.005076              0.695652   
2                           0.001982              0.339130   
3                           0.000169              0.584980   
4                           0.001822              0.623188   
...                              ...                   ...   
3737                        0.001839              0.434783   
3738                        0.002038              0.652174   
3739                        0.004472              0.604391   
3740                        0.001339              0.629156   
3741                        0.002047              0.571146   

      Avg_Transaction_Day  Avg_Transaction_Month  Avg_Transaction_Year  \
0                0.666667               0.909091              0.000000   
1                0.666667               0.909091              0.000000   
2                0.500000               0.909091              0.000000   
3                0.457576               0.694215              0.272727   
4                0.000000               0.090909              1.000000   
...                   ...                    ...                   ...   
3737             0.233333               0.090909              1.000000   
3738             0.566667               0.000000              1.000000   
3739             0.582508               0.382538              0.623762   
3740             0.200000               1.000000              0.000000   
3741             0.498485               0.033058              1.000000   

      RFMS_Score Risk_Label  Recency_Bin Frequency_Bin Volume_Bin  Recency_WoE  
0       0.369991       Good     Very Old      Very Low   Very Low     1.000000  
1       0.369991       Good     Very Old      Very Low   Very Low     1.000000  
2       0.384011       Good     Very Old           Low        Low     1.000000  
3       0.211280       Good          Old          High        Low     1.000000  
4       0.172534        Bad       Recent           Low        Low     0.028200  
...          ...        ...          ...           ...        ...          ...  
3737    0.153303        Bad  Very Recent           Low        Low     0.002047  
3738    0.210934       Good          Old           Low        Low     1.000000  
3739    0.148816        Bad  Very Recent     Very High  Very High     0.002047  
3740    0.327210       Good     Very Old 

In [96]:
m.logistic_regression_model_evaluation(features)

2024-10-12 02:45:15,834 - INFO - logistic regresion Model_evaluation.....
2024-10-12 02:45:15,834 - INFO - logistic regresion Model_evaluation.....
Accuracy: 99.07%
Precision: 1.00
Recall: 0.98
F1 Score: 0.99
ROC-AUC: 1.00
Confusion Matrix:
[[378   0]
 [  7 364]]


In [97]:
m.random_forest_model_evaluation(features)

2024-10-12 02:45:15,993 - INFO - random forest Model_evaluation.....
2024-10-12 02:45:15,993 - INFO - random forest Model_evaluation.....
Accuracy: 0.9879839786381842
Precision: 0.994535519125683
Recall: 0.9811320754716981
F1 Score: 0.9877883310719131
ROC-AUC: 0.9939852251173006
Confusion Matrix: [[376   2]
 [  7 364]]


The logistic regression model evaluation results based on the binned features are as follows:

    Accuracy: 99.07% – The model correctly predicts 99.07% of the total observations.
    Precision: 100% – All positive predictions (classified as "Good") are correct.
    Recall (Sensitivity): 98% – The model identifies 98% of the actual "Good" cases.
    F1 Score: 99% – A balanced measure between precision and recall.
    ROC-AUC: 100% – The model's ability to distinguish between "Good" and "Bad" classes is excellent.

so compared to the random forest we got logestic_regression the best although the results are near

so we are going to fit the logistic regression Model

In [98]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, precision_score, recall_score, f1_score, roc_auc_score


In [99]:
# Use only WoE transformed features and Risk_Label for modeling
X = features[['Recency_WoE', 'Transaction_Frequency', 'Total_Transaction_Volume']]
y = features['Risk_Label'].map({'Good': 1, 'Bad': 0})  # Encode labels as 1 for Good, 0 for Bad

    # Split into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Logistic Regression model
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)

    # Make predictions
y_pred = logistic_model.predict(X_test)

In [100]:
m.fitting_model

<function Model.fitting_model(X, y)>

In [101]:

# Save the model
with open('logistic_regression_model.pkl', 'wb') as model_file:
    pickle.dump(final_model, model_file)


In [104]:
# Load the model
with open('logistic_regression_model.pkl', 'rb') as model_file:
    loaded_model = pickle.load(model_file)

# Predicting on new data (replace `new_data` with actual new data)
new_data_predictions = loaded_model.predict(new_data)


In [103]:
# Load the model
with open('logistic_regression_model.pkl', 'rb') as model_file:
    loaded_model = pickle.load(model_file)

# Predicting on new data (replace `new_data` with actual new data)
new_data_predictions = loaded_model.predict(new_data)


In [106]:
new_data

Recency_WoE  Transaction_Frequency  Total_Transaction_Volume
3450     0.002047               0.009535                  0.557658
1114     0.002047               0.000244                  0.557103
351      0.002047               0.000733                  0.556991
1983     0.002047               0.010513                  0.557213
2321     0.028200               0.000000                  0.556965

In [105]:
# Print feature importance
feature_importance = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': final_model.coef_[0]
})

print(feature_importance.sort_values(by='Coefficient', ascending=False))


                    Feature  Coefficient
0               Recency_WoE     9.385466
1     Transaction_Frequency     0.958783
2  Total_Transaction_Volume     0.948137
